### PR TITLE
[CFM-1210] -Only GO/GA can add/edit wiki pages by default

### PR DESCRIPTION
**Pull request by ybaras**
https://issuetracker.amplexor.com/jira/browse/CFM-1210

### CODE REVIEW

**Comment by ybaras** [comment]
GO can add and edit WIKI pages:

can

Others cannot:

cant

**Comment by ordavidil** [comment]
@ybaras 
The change should be over here:
https://github.com/capacity4dev/capacity4more/blob/develop/project/profiles/capacity4more/capacity4more.profile#L135

File: project/profiles/capacity4more/capacity4more.profile [hunk 1]
```diff
@@ -149,8 +148,19 @@ function capacity4more_setup_set_og_permissions() {
...
+  $content_types = array(
+    'wiki_page',
+  );
 
+  $permissions = array();
+  foreach ($content_types as $content_type) {
+    $permissions = array_merge($permissions, array(
+      "create $content_type content",
+      "update own $content_type content",
+      "delete own $content_type content",
```
Comment (ordavidil) [inline project/profiles/capacity4more/capacity4more.profile]: @ybaras 
Please remove the _delete_ permission since this is not part of the request.

selection_212

**Comment by soniCaH** [comment]
Regular GM can still add wiki pages to a group.

**Comment by ordavidil** [comment]
@ybaras 
add screenshot for GM please

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php [hunk 2]
```diff
@@ -87,8 +87,15 @@ public function buildEntityFieldQuery(
     }
 
     $user_groups = og_get_groups_by_user(NULL, $group_type);
-    $user_groups = $user_groups ? $user_groups : array();
+    $user_groups = $user_groups ?: array();
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php]: @ybaras 
you can revert the change of this line since it does the same thing but it is less readable ...

**Comment by ordavidil** [comment]
@ybaras 
many tests failed, please check it.

**Comment by ybaras** [comment]
@ordavidil Behat fails because the quickpost forms are throwing an error (AngularJS problem).
Adding content through the regular form works OK, including permission tests.

File: project/profiles/capacity4more/behat/features/bootstrap/FeatureContext/Discussion.php [hunk 3]
```diff
@@ -62,12 +62,12 @@ public function iShouldSeeTheDiscussionDetailPage() {
    * @Given /^a discussion "([^"]*)" in group "([^"]*)" is created$/
    */
   public function aDiscussionInGroupIsCreated($title, $group_title) {
+    $machine_readable_group = strtolower(preg_replace('@[^a-zA-Z0-9_]+@', '-', trim($group_title)));
```
Comment (ordavidil) [inline project/profiles/capacity4more/behat/features/bootstrap/FeatureContext/Discussion.php]: @ybaras 
Please add a fine comment for this one, thanks.

File: project/profiles/capacity4more/behat/features/bootstrap/FeatureContext/Node.php [hunk 4]
```diff
@@ -62,12 +62,13 @@ public function iShouldNotBeAllowedToCreateA($type) {
   }
 
   /**
-   * @Then /^I should be allowed to create a "([^"]*)"$/
+   * @Then /^I should be allowed to create a "([^"]*)" in group "([^"]*)"$/
    */
-  public function iShouldBeAllowedToCreateA($type) {
+  public function iShouldBeAllowedToCreateA($type, $group) {
+    $machine_readable_group = strtolower(preg_replace('@[^a-zA-Z0-9_]+@', '-', trim($group)));
```
Comment (ordavidil) [inline project/profiles/capacity4more/behat/features/bootstrap/FeatureContext/Node.php]: @ybaras 
It seems the same as the other one,  maybe it'll be better to take out to an external function ?

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php [hunk 5]
```diff
@@ -90,6 +90,13 @@ public function buildEntityFieldQuery(
     $user_groups = og_get_groups_by_user(NULL, $group_type);
     $user_groups = $user_groups ? $user_groups : array();
 
+    $current_gid = og_context();
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php]: @ybaras 
`og_context()` returns the `$group` object, please rename the variable.

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php [hunk 6]
```diff
@@ -90,6 +90,13 @@ public function buildEntityFieldQuery(
...
+    $node_type = $this->instance['bundle'];
+    if (!og_user_access($group_type, $current_gid['gid'], "create $node_type content")) {
+      // User does not have permission, falsify the query.
+      $query->propertyCondition($entity_info['entity keys']['id'], -1, '=');
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php]: @ybaras 
What's the `-1` stands for ?

Comment (ybaras) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php]: An invalid value to test the id field against.

**Comment by ybaras** [comment]
test-record

**Comment by ybaras** [comment]
alb

File: project/profiles/capacity4more/behat/features/bootstrap/FeatureContext/Node.php [hunk 7]
```diff
@@ -221,4 +220,15 @@ public function iUpdateAWithTitleInTheGroupWithNewTitleAfter($type, $title, $new
...
     node_save($node);
   }
+
+  /**
+   * Convert a human readable title to a machine readable name.
+   *
+   * @return string
+   */
+  public function humanToMachineReadable($title, $delimiter) {
+
```
Comment (ordavidil) [inline project/profiles/capacity4more/behat/features/bootstrap/FeatureContext/Node.php]: @ybaras 
please an elaborated comment about this one (especially the regex).

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php [hunk 8]
```diff
@@ -90,6 +90,13 @@ public function buildEntityFieldQuery(
     $user_groups = og_get_groups_by_user(NULL, $group_type);
     $user_groups = $user_groups ? $user_groups : array();
 
+    $group_obj = og_context();
+    $node_type = $this->instance['bundle'];
+    if (!og_user_access($group_type, $group_obj['gid'], "create $node_type content")) {
+      // User does not have permission, falsify the query.
+      $query->propertyCondition($entity_info['entity keys']['id'], -1, '=');
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php]: @ybaras 
What's the `-1` stands for?

Comment (ybaras) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php]: An invalid ID. Sending a false query when the user is not authorized. This method is used 3 times in the file.

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php]: @ybaras 
so please write it down on the code.
these kind of numbers should not live in the code without a good explanation.
I would create a constant for that on the beginning of the code and replace it all.

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php [hunk 9]
```diff
@@ -90,6 +93,13 @@ public function buildEntityFieldQuery(
     $user_groups = og_get_groups_by_user(NULL, $group_type);
     $user_groups = $user_groups ? $user_groups : array();
 
+    $group_obj = og_context();
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php]: Can we use just `$group` in here ?

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php [hunk 10]
```diff
@@ -10,6 +10,9 @@
  */
 class C4MOgSelectionHandler extends OgSelectionHandler {
 
+  // Defines a false, invalid ID, to deny access to unauthorized users.
+  const FALSEID = -1;
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/plugins/entityreference/selection/C4MOgSelectionHandler.class.php]: maybe
```php
const FALSE_ID = -1;
```
?

File: project/profiles/capacity4more/behat/features/bootstrap/FeatureContext/Node.php (hunk 7 above)
Comment (ordavidil) [inline project/profiles/capacity4more/behat/features/bootstrap/FeatureContext/Node.php]: @ybaras 
remove the empty line.
